File: xim/resource/table/ItemModelTable.kt
package xim.resource.table

import xim.poc.ItemModelSlot
import xim.poc.browser.DatLoader
import xim.poc.game.configuration.constants.armorKubiraMeikogai_26959
import xim.poc.game.configuration.constants.armorMakoraMeikogai_26961
import xim.resource.ByteReader
import xim.resource.InventoryItemInfo
import xim.resource.table.AdditionalModelMapping.additionalMappings

// https://github.com/LandSandBoat/server/blob/base/sql/item_equipment.sql
object ItemModelTable: LoadableResource {

    private lateinit var table: ByteReader
    private var preloaded = false

    override fun preload() {
        if (preloaded) { return }
        preloaded = true
        loadTable()
    }

    override fun isFullyLoaded() : Boolean {
        return this::table.isInitialized
    }

    operator fun get(item: InventoryItemInfo?) : Int {
        return getModelId(item)
    }

    fun getForcedMatches(slot: ItemModelSlot, modelId: Int): Set<ItemModelSlot> {
        return ItemModelMatcher.itemModelOverrides[modelId]?.get(slot) ?: emptySet()
    }

    private fun getModelId(item: InventoryItemInfo?) : Int {
        if (item == null) { return 0 }
        return getModelId(item.itemId)
    }

    private fun getModelId(itemId: Int) : Int {
        val additional = additionalMappings[itemId]
        if (additional != null) { return additional }

        table.position = itemId * 2
        return table.next16()
    }

    private fun loadTable() {
        DatLoader.load("landsandboat/ItemModelTable.DAT").onReady { table = it.getAsBytes() }
    }

}

private object ItemModelMatcher {

    private val swimwearOverride = mapOf(
        ItemModelSlot.Body to setOf(ItemModelSlot.Hands),
        ItemModelSlot.Legs to setOf(ItemModelSlot.Feet),
    )

    val itemModelOverrides = mapOf(
        495 to swimwearOverride,
        496 to swimwearOverride,
    )

}

private object AdditionalModelMapping {

    val additionalMappings = mapOf(
        23871 to 495,   // Hebenus Gilet
        23872 to 495,   // Hebenus Boxers
        23873 to 496,   // Hebenus Top
        23874 to 496,   // Hebenus Shorts
        26959 to 303,   // Kubira Meikogai
        26961 to 209,   // Makora Meikogai
        26962 to 51,    // Enforcer's Harness
    )

}

File: lib/resource/table/itemModelTable.ts
import { ByteReader } from '../byteReader'
import type { LoadableResource } from './loadableResource'
import type { InventoryItemInfo, ItemModelSlot } from './tableTypes'
import { ItemModelSlot as ItemModelSlots } from './tableTypes'

const swimwearOverride = new Map<ItemModelSlot, ReadonlySet<ItemModelSlot>>([
  [ItemModelSlots.Body, new Set([ItemModelSlots.Hands])],
  [ItemModelSlots.Legs, new Set([ItemModelSlots.Feet])],
])

const itemModelOverrides = new Map<number, ReadonlyMap<ItemModelSlot, ReadonlySet<ItemModelSlot>>>([
  [495, swimwearOverride],
  [496, swimwearOverride],
])

const additionalMappings = new Map<number, number>([
  [23871, 495],
  [23872, 495],
  [23873, 496],
  [23874, 496],
  [26959, 303],
  [26961, 209],
  [26962, 51],
])

export class ItemModelTable implements LoadableResource {
  private readonly loadTableBytes: () => Promise<Uint8Array>
  private table: ByteReader | null = null
  private preloaded = false

  constructor(loadTableBytes: () => Promise<Uint8Array>) {
    this.loadTableBytes = loadTableBytes
  }

  async preload(): Promise<void> {
    if (this.preloaded) {
      return
    }

    this.preloaded = true
    this.table = new ByteReader(await this.loadTableBytes())
  }

  isFullyLoaded(): boolean {
    return this.table !== null
  }

  getModelId(item: InventoryItemInfo | null): number {
    if (item === null) {
      return 0
    }

    return this.getModelIdFromItemId(item.itemId)
  }

  getModelIdFromItemId(itemId: number): number {
    const mapped = additionalMappings.get(itemId)
    if (mapped !== undefined) {
      return mapped
    }

    if (this.table === null) {
      throw new Error('ItemModelTable must be preloaded before use')
    }

    this.table.position = itemId * 2
    return this.table.next16()
  }

  getForcedMatches(slot: ItemModelSlot, modelId: number): ReadonlySet<ItemModelSlot> {
    return itemModelOverrides.get(modelId)?.get(slot) ?? new Set<ItemModelSlot>()
  }
}
